### ISSUE TITLE
insights: commit index has duplicate revisions across different repos

### ISSUE THREAD

**Issue by coury-clark** [sourcegraph/sourcegraph#29922]
While debugging an issue with a Code Insight, I ran across a problem where the commit index had some duplicate revisions across multiple repositories. In all of the duplicates they were paired (repository X, Y shared a revision). I observed this both locally and on k8s-dogfood.

The result was a search that was queued with a revision that did not exist in the repository, and a broken insight. Slack [thread](https://sourcegraph.slack.com/archives/C02KE7E9GMR/p1641583095002800)

I resolved this by regenerating the index, but we need to understand the root cause and fix this (and likely regenerate all customer indices). Some possible ideas:
1. Somehow `gitserver` has a bug and is returning bad data
2. Something is getting cached in the indexer and writing the revision to the wrong repository

**Comment by coury-clark** [comment]
There are currently 95,980 revisions on k8s-dogfood that have at least 1 duplicate in a separate repository. This indicates there is a significant problem in the indexer. Whether or not this problem has always existed is not currently clear.

**Comment by coury-clark** [comment]
To summarize work that is spread out in disparate places:
1. Revisions that were placed in the commit index no longer existed on *any* repository in the Sourcegraph installation (including the ones it was associated with in the commit index)
2. Two other bugs were found in the commit index that were causing related problems https://github.com/sourcegraph/sourcegraph/issues/30434 https://github.com/sourcegraph/sourcegraph/issues/30433 and were fixed
3. We have modified the backfiller to always look up revisions at at time, rather than use the revisions in the commit index to work around this problem. This has the effect that if there is somehow revisions missing we will still produce bad series, but if extra commits exist all we are doing is formulating extra work but will yield the same result.

For now I am closing this issue and opening a new issue to tackle the first point above at some point in the future.
